src/libsystemd/sd-device/sd-device.c

```diff
@@ -1877,28 +1877,29 @@ static int device_sysattrs_read_all_internal(sd_device *device, const char *subd
                 return r;
 
         if (subdir) {
-                _cleanup_free_ char *p = NULL;
-
-                p = path_join(syspath, subdir, "uevent");
-                if (!p)
-                        return -ENOMEM;
-
-                if (access(p, F_OK) >= 0)
-                        /* this is a child device, skipping */
-                        return 0;
-                if (errno != ENOENT) {
-                        log_device_debug_errno(device, errno, "sd-device: Failed to stat %s, ignoring subdir: %m", p);
-                        return 0;
-                }
-
                 path_dir = path_join(syspath, subdir);
                 if (!path_dir)
                         return -ENOMEM;
         }
 
         dir = opendir(path_dir ?: syspath);
-        if (!dir)
+        if (!dir) {
+                if (errno == ENOENT && subdir)
+                        return 0; /* Maybe, this is a child device, and is already removed. */
+
                 return -errno;
+        }
+
+        if (subdir) {
+                if (faccessat(dirfd(dir), "uevent", F_OK, 0) >= 0)
```

Comment:
Shouldn't this be `subdir`?

Reply:
existance of "uevent" file indicates a proper device in sysfs. If it's absent then a dir is just more attributes of the original device. Hence the check here: if a `subdir` is specified we check if `uevent` exists in it, and if so, then we skip the dir, because in that case it would be attributes of some other device.

Reply:
Right. Here, `"uevent"` is correct, as also briefly commented in the code, this check is for detecting child device.